tasks/fullStack/app/controllers/reviews_controller.rb

```diff
@@ -1,29 +1,45 @@
 class ReviewsController < ApplicationController
+  before_action :set_product, only: [:new]
 
   DEFAULT_TAGS = ['default']
 
   def index
-    if params[:shop_id].present? && Shop.where("id = #{params[:shop_id]}").present?
+    params[:shop_id] ||= 1
+    @shop = Shop.find(params[:shop_id])
+
+    if @shop.present?
       params[:per_page] ||= 10
       offset = params[:page].to_i * params[:per_page]
 
+      @shops = Shop.where.not(id: params[:shop_id]).limit(5)
+
       @data = []
-      products = Product.where("shop_id = #{params[:shop_id]}").sort_by(&:created_at)[offset..(offset + params[:per_page])]
+      products = Product.where("shop_id = #{params[:shop_id]}").limit(5).sort_by(&:created_at)[offset..(offset + params[:per_page])]
       products.each do |product|
-        reviews = product.reviews.sort_by(&:created_at)[offset..(offset + params[:per_page])]
+        reviews = product.reviews.limit(5).sort_by(&:created_at)[offset..(offset + params[:per_page])]
```

Comment:
Can you identify an anti-pattern that has negative performance impact here and suggest ways to improve it

Reply:
This query suffers from the N+1 query problem. The best way to optimise this would be to eager load the reviews alongside the products in a single query. Additonally, I also believe it is more efficient to sort the records at the database level using `.order`. So I believe the updated query would look something like this:

`products = Product.includes(:reviews)
  .where(shop_id: params[:shop_id])
  .order(created_at: :desc)
`

For readability `.map` can be used instead of `.each` to store the result of the loop to `@data` like so:

`
@data = products.map do |product|
  { product: product, reviews: product.reviews.limit(5).sort_by(&:created_at) }
end
`